src/common/device_helpers.cuh

```diff
@@ -391,10 +408,16 @@ struct XGBDefaultDeviceAllocatorImpl : thrust::device_malloc_allocator<T> {
     GlobalMemoryLogger().RegisterDeallocation(ptr.get(), n * sizeof(T));
     return SuperT::deallocate(ptr, n);
   }
+#if defined(XGBOOST_USE_RMM) && XGBOOST_USE_RMM == 1
+  using cuda_mr = rmm::mr::cuda_memory_resource;
+  using pool_mr = rmm::mr::pool_memory_resource<cuda_mr>;
+  XGBDefaultDeviceAllocatorImpl() : SuperT(new pool_mr(new cuda_mr), cudaStream_t{0}) {}
```

Comment:
I see that the cuda_mr and pool_mr are created here, but they don't appear to be deleted anywhere in this PR. This would be a pretty significant leak.

More importantly, if a goal is to use use RMM so that workflows that use RAPIDS libraries in the same process as XGBoost can share a memory pool, then you would need a way to register an existing RMM memory resource rather than creating a new one. A common use case would be to do ETL in cuDF (which creates the RMM resource) and then use the same RMM resource in XGBoost.  So perhaps you need an option where XGBoost would all `rmm::mr::get_default_resource()` to get the previously configured MR instead of creating a new one?

Reply:
I now see that @RAMitchell already suggested this.

Reply:
@harrism Indeed, using `rmm::mr::get_default_resource()` fixed the crash. The tests now pass.

Reply:
No worries!

Reply:
> @harrism Indeed, using `rmm::mr::get_default_resource()` fixed the crash. The tests now pass. Thank you!

So it's not too surprising that the tests would pass now as the initial default resource is just a `cuda_memory_resource` that uses normal cudaMalloc/cudaFree for allocation.

It's still a little worrying to me that using the pool causes OOMs.

Reply:
@jrhemstad 

> It's still a little worrying to me that using the pool causes OOMs.

No, the OOM was due to having multiple copies of `pool_mr` being constructed with `new`. I tried setting the default resource with `pool_mr` using `rmm::mr::set_default_resource()`, and the tests still pass. See my latest commit.

Reply:
@jrhemstad Oops, I found out that instead of OOM, I'm seeing a segfault now, when a Thrust device vector backed with a pool allocated buffer grows:
```
[ RUN      ] BitField.GPUSet

Thread 1 "testxgboost" received signal SIGSEGV, Segmentation fault.
0x000055555583ffce in thrust::detail::vector_base<unsigned long, dh::detail::XGBDefaultDeviceAllocatorImpl<unsigned long> >::append(unsigned long) ()
(gdb) git backtracQuit
(gdb) backtrace
#0  0x000055555583ffce in thrust::detail::vector_base<unsigned long, dh::detail::XGBDefaultDeviceAllocatorImpl<unsigned long> >::append(unsigned long) ()
#1  0x0000555555ab1356 in xgboost::BitField_GPUSet_Test::TestBody() ()
#2  0x0000555555bb3c67 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
#3  0x0000555555ba3fb3 in testing::Test::Run() ()
#4  0x0000555555ba4158 in testing::TestInfo::Run() ()
#5  0x0000555555ba43b5 in testing::TestSuite::Run() ()
#6  0x0000555555baa5c0 in testing::internal::UnitTestImpl::RunAllTests() ()
#7  0x0000555555baa825 in testing::UnitTest::Run() ()
#8  0x00005555555db70d in main ()
```

Reply:
That's using `pool_memory_resource`? How about if using `cnmem_memory_resource`?

Reply:
I'm actually using `cnmem_memory_resource()`.

Reply:
Hm..., in that case, how about if you use `pool_memory_resource`?

Reply:
I fixed the segfault by allocating `pool_mr` in the heap rather than in the stack. Some unit tests in our Span implementation are failing, will take a look.